Mallory/align

wangmallory opened the pull request:
Adding to case_studies.coadds:
- align_single_exposures which takes original image and dithered images to return interpolated grid (n x 1 x slen x slen) 
- signal_noise_ratio which calculates signal to noise ratio 

And in tests.case_studies.coadds
- test_align_single_exposures

ismael-mendoza commented on case_studies/coadds/align_single_exposures.py:
```diff
@@ -0,0 +1,39 @@
+from torch import Tensor
+from bliss.models.decoder import get_mgrid
+
+def align_single_exposures(img0, img1, slen, pixel_scale, dithers):
```
It would be great if you could add documentation to this function, specifying what each variable is and what are the dimensions of the input tensors

ismael-mendoza commented on case_studies/coadds/align_single_exposures.py:
```diff
@@ -0,0 +1,39 @@
+
+    iplots = []
+    input = torch.tensor(images[:]).reshape(len(dithers),1,slen,slen).float()
+    grids = grid.reshape(len(dithers),1,slen*slen,2).float()
```
`slen*slen` -> `slen**2`

ismael-mendoza commented on case_studies/coadds/align_single_exposures.py:
```diff
@@ -0,0 +1,39 @@
+    
+    tenplot = torch.tensor(iplots[:][0])
+    crop_images = []
+    for i in range(tenplot.shape[0]):
+        im = np.array(tenplot[i].reshape(slen,slen))
+        crop_im = im[1:slen-1, 1:slen-1]
+        crop_images.append(crop_im)
```
do you need this `for` loop?

ismael-mendoza commented on case_studies/coadds/signal_noise_ratio.py:
```diff
@@ -0,0 +1,19 @@
+def n_s_ratio(img, background):
+    """Calculate Signal-to-Noise Ratio.
+    Args:
+        img: Tensor of shape `(N x C x H x W)` where N is the batch_size, C is the \
```
you don't need the `\` at the end I don't think (since the documentation uses tripe quotes.

ismael-mendoza commented on case_studies/coadds/signal_noise_ratio.py:
```diff
@@ -0,0 +1,19 @@
+    """Calculate Signal-to-Noise Ratio.
+    Args:
+        img: Tensor of shape `(N x C x H x W)` where N is the batch_size, C is the \
+            number of band, H is height and W is weight, containing image data.
```
```suggestion
            number of band, H is height and W is weight, containing image data (with background).
```

ismael-mendoza replied:
be careful to use this function only when the image already has background added

ismael-mendoza commented on tests/case_studies/coadds/test_align_single_exposures.py:
```diff
@@ -0,0 +1,18 @@
+    dithers = [((-0.5 - 0.5) * torch.rand((2,)) + 0.5).numpy() for x in range(10)]
+    img = []
+    for i in dithers:
+        wcs1 = galsim.OffsetWCS(scale = 0.393, origin = galsim.PositionD(i))
```
why do you need this `wcs1`?

ismael-mendoza commented on case_studies/coadds/align_single_exposures.py:
```diff
@@ -0,0 +1,39 @@
+    images = img1
+
+    sgrid = (get_mgrid(slen) - (-1))/(1-(-1)) * (slen)
+    grid_x = wcs0.xyTouv(np.array(sgrid.reshape(slen*slen,2)[:,0]), np.array(sgrid.reshape(slen*slen,2)[:,1]))[0]
+    grid_y = wcs0.xyTouv(np.array(sgrid.reshape(slen*slen,2)[:,0]), np.array(sgrid.reshape(slen*slen,2)[:,1]))[1]
```
do you need the `np.array` here? perhaps better to convert `sgrid` to an array before hand? (if it's a tensor you can call it `sgrid_np` after casting it)

ismael-mendoza commented on case_studies/coadds/align_single_exposures.py:
```diff
@@ -0,0 +1,39 @@
+# galsim_decoder changes for coadds
```
could you add a minimal notebook showing the `align_single_exposures` functions work with a couple of images?  You probalby already showed that in the meeting sometime and just have to remove some extra stuff. Thank you :)

ismael-mendoza commented on case_studies/coadds/align_single_exposures.py:
```diff
@@ -0,0 +1,49 @@
+    iplots = []
+    input = torch.tensor(images[:]).reshape(len(dithers), 1, slen, slen).float()
+    grids = grid.reshape(len(dithers), 1, slen * slen, 2).float()
+    iplots.append(F.grid_sample(input, grids, align_corners=False))
```
`interp_images`

ismael-mendoza commented on case_studies/coadds/signal_noise_ratio.py:
```diff
@@ -0,0 +1,19 @@
+import torch
+
+
+def snr(img, background):
```
`img` -> `img_no_background`

ismael-mendoza commented on tests/case_studies/coadds/test_align_single_exposures.py:
```diff
@@ -0,0 +1,20 @@
+        im = g0.drawImage(nx=slen, ny=slen, scale=pixel_scale, offset=i, bandpass=None)
+        im = im.array
+        img.append(im)
+    align_single_exposures(img0, img, slen, dithers)
```
assert shape

wangmallory commented:
@ismael-mendoza I've made all of the updates we discussed yesterday. For snr, I changed the calculation to average signal / standard deviation signal because that made the most sense to me per the readings (and removed background since no noise has been made). But let me know if that's not the best equation.

wangmallory commented:
@ismael-mendoza here's the updated snr

ismael-mendoza commented on case_studies/coadds/align_single_exposures.py:
```diff
@@ -0,0 +1,48 @@
+
+
+def align_single_exposures(img0, images, slen, dithers, scale=0.393):
+    """Aligns multiple single exposure images that are dithered by some sub-pixel amount.
+    Args:
```
```suggestion
    """Aligns multiple single exposure images that are dithered by some sub-pixel amount.
    
    Args:
```

ismael-mendoza commented on case_studies/coadds/align_single_exposures.py:
```diff
@@ -0,0 +1,49 @@
+            C is the number of band, H is height and W is weight, containing image data.
+        slen: size of images (H and W)
+        dithers: List of pairs of sub-pixel amounts that img0 is shifted by in x and y directions
+        scale: pixel_scale
+    Returns:
```
```suggestion
        scale: pixel_scale
        
    Returns:
```

ismael-mendoza commented:
@wangmallory just made a couple of small changes, thanks for this PR! I think it's ready to be merged, please just review my recent changes and if you agree please merge!
